dsl/src/main/scala/com.crobox.clickhouse/dsl/column/HigherOrderFunctions.scala

```diff
@@ -3,32 +3,240 @@ package com.crobox.clickhouse.dsl.column
+      _func3: Option[(TableColumn[I], TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _arrays: ArrayColMagnet[_ <: Iterable[I]]*
+  ) extends HigherOrderFunction[I, Boolean, Iterable[I]](_func1, _func2, _func3, _arrays: _*)
+  case class ArrayExists[I](
+      _func1: Option[TableColumn[I] => ExpressionColumn[Boolean]],
+      _func2: Option[(TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _func3: Option[(TableColumn[I], TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _arrays: ArrayColMagnet[_ <: Iterable[I]]*
+  ) extends HigherOrderFunction[I, Boolean, Iterable[I]](_func1, _func2, _func3, _arrays: _*)
+  case class ArrayFilter[I](
+      _func1: Option[TableColumn[I] => ExpressionColumn[Boolean]],
+      _func2: Option[(TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _func3: Option[(TableColumn[I], TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _arrays: ArrayColMagnet[_ <: Iterable[I]]*
+  ) extends HigherOrderFunction[I, Boolean, Iterable[I]](_func1, _func2, _func3, _arrays: _*)
+  case class ArrayFirst[I](
+      _func1: Option[TableColumn[I] => ExpressionColumn[Boolean]],
+      _func2: Option[(TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _func3: Option[(TableColumn[I], TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _arrays: ArrayColMagnet[_ <: Iterable[I]]*
+  ) extends HigherOrderFunction[I, Boolean, Iterable[I]](_func1, _func2, _func3, _arrays: _*)
+  case class ArrayFirstIndex[I](
+      _func1: Option[TableColumn[I] => ExpressionColumn[Boolean]],
+      _func2: Option[(TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _func3: Option[(TableColumn[I], TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean]],
+      _arrays: ArrayColMagnet[_ <: Iterable[I]]*
+  ) extends HigherOrderFunction[I, Boolean, Iterable[I]](_func1, _func2, _func3, _arrays: _*)
+
+  def arrayAll[I](func: TableColumn[I] => ExpressionColumn[Boolean],
+                  array: ArrayColMagnet[_ <: Iterable[I]]): ArrayAll[I, Boolean] =
+    ArrayAll(Option(func), None, None, array)
+
+  def arrayAll2[I](func: (TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean],
+                   array1: ArrayColMagnet[_ <: Iterable[I]],
+                   array2: ArrayColMagnet[_ <: Iterable[I]]): ArrayAll[I, Boolean] =
+    ArrayAll(None, Option(func), None, array1, array2)
+
+  def arrayAll3[I](func: (TableColumn[I], TableColumn[I], TableColumn[I]) => ExpressionColumn[Boolean],
+                   array1: ArrayColMagnet[_ <: Iterable[I]],
+                   array2: ArrayColMagnet[_ <: Iterable[I]],
+                   array3: ArrayColMagnet[_ <: Iterable[I]]): ArrayAll[I, Boolean] =
+    ArrayAll(None, None, Option(func), array1, array2, array3)
+
+  def arrayCount[I](func: Option[TableColumn[I] => ExpressionColumn[Boolean]],
```

Comment:
Should we allow this to be `func: Option` since f.e. `arrayAll` and `arrayExists` have no such thing